filters/imagefilter.go

```diff
@@ -43,65 +41,74 @@ func init() {
 
 type ImageFilter interface {
 	CreateOptions(image *bimg.Image) (*bimg.Options, error)
+	CanBeMerged(other *bimg.Options, self *bimg.Options) bool
+	Merge(other *bimg.Options, self *bimg.Options) *bimg.Options
 }
 
 func HandleImageResponse(ctx filters.FilterContext, f ImageFilter) {
-	rsp := ctx.Response()
-
-	defer rsp.Body.Close()
-
-	rsp.Header.Del("Content-Length")
-
-	r, err := handleImageTransform(rsp.Body, f)
+	image := ctx.StateBag()[SkropImage].(*bimg.Image)
 
+	opt, err := f.CreateOptions(image)
 	if err != nil {
-		log.Error("failed to process image ", err.Error())
+		log.Error("Failed to create options ", err.Error())
 		ctx.Serve(&http.Response{
-			StatusCode:http.StatusInternalServerError,
-			Body: ioutil.NopCloser(bytes.NewBufferString(err.Error())),
+			StatusCode: http.StatusInternalServerError,
+			Body:       ioutil.NopCloser(bytes.NewBufferString(err.Error())),
 		})
 		return
 	}
 
-	rsp.Body = ioutil.NopCloser(r)
-}
+	opts := ctx.StateBag()[SkropOptions].(*bimg.Options)
 
-func handleImageTransform(r io.Reader, f ImageFilter) (io.Reader, error) {
-	var err error
-	imageBytes, err := ioutil.ReadAll(r)
+	if f.CanBeMerged(opts, opt) {
+		ctx.StateBag()[SkropOptions] = f.Merge(opts, opt)
+		log.Debug("Filter ", f, " merged in ", ctx.StateBag()[SkropOptions])
+		return
+	}
 
+	//transform the image
+	buf, err := transformImage(image, opts)
```

Comment:
i agree with @chandu188 in this case, because when we say "a filter returns with a new bimg.Option" we could replace that with "append an instruction to a slice that will cause the defined 'step' return with the bimg.Option during applying the instructions". Or am I misunderstanding something?